Asset key to node handle mapping

clairelin135 opened the pull request:
This PR generates a mapping from asset key to the (non-asset) nodes required to run to generate the asset. This implementation works only for the new implementation of asset jobs, where asset definition information exists at the top level of the job, and not for the old implementation where asset keys exist on output definitions.

Example cases this PR handles:
- Every node in the job is an `@asset`:
    - Each asset key of each node is mapped to a set containing the node
- Graph contains ops in sequence and last op output is an asset
    - Returns map of asset key -> set containing all op node handles
- Graph contains 2 disconnected ops and each output is a different asset
    - Returns map of each asset key -> set containing the singular op that outputted the asset
- Graph has op `o1` that outputs asset `a1` and op `o2` that has input `o1` but no output
    - Returns map of `a1` to `o1`
- Nested case: graph `g1` with op `o1` is an input to outer graph `g2` that returns `o1` without mutating
    - Returns map of `o1` asset key -> set containing node handle of `o1`
- Internal asset deps
    - Job contains multi_asset `two_assets` with outputs `o1` and `o2`, where `o2` depends exclusively on `o1`
        - Returns map where assets for `o1` and `o2` are both mapped to `two_assets`
    - Job contains graph where op `o1` is an input to op `o2`, which outputs asset `a1` and `a2`. `a2` has defined `internal_asset_deps` to only `a1`
        - Returns `a1` mapped to `o1` and `o2`
        - Returns `a2` mapped to `o2`

Not independently mergeable until base branch has been merged.

sryza reviewed (commented):
This PR is quite the software engineering interview question!

I think there might be opportunity to do this more efficiently. Because we're dealing with graphs, there might be multiple paths to the same op handle, even from the same asset key. Are we able to treat this like a DFS and avoid traversing the same nodes multiple times. Or maybe I missed it and we're already doing that?

I also left a few comments about code readability. I think it's worth taking extra care in situations like these when the code is so complex.

clairelin135 commented on python_modules/dagster/dagster/core/definitions/assets_info.py:
```diff
@@ -64,6 +67,136 @@ def __new__(
+    graph_def: GraphDefinition, assets_defs_by_node_handle: Mapping[NodeHandle, "AssetsDefinition"]
+) -> Mapping[AssetKey, Set[NodeHandle]]:
+    non_asset_inputs_by_node_handle: Dict[NodeHandle, Sequence[Tuple[str, NodeHandle]]] = {}
+    outputs_by_graph_handle: Dict[NodeHandle, Dict[str, NodeHandle]] = {}
```
`outputs_by_graph_handle`: A mapping of every graph node handle to a dictionary with each out name as a key and a tuple containing (op output name, op node handle) as the value representing the op output within the graph

`non_asset_inputs_by_node_handle`: A mapping of all node handles to all upstream node handles that are not assets. Each key is a node handle with tuple value (node_output_name, output_node) representing the output node

clairelin135 commented:
@sryza Thanks for the feedback, I updated this to add a lot more docstrings and inline comments containing implementation details. I also added memoization so that the graph traversal will stop traversal at already visited nodes. I also addressed the function renames and type annotations.

sryza commented on python_modules/dagster/dagster/core/definitions/assets_info.py:
```diff
@@ -64,6 +67,188 @@ def __new__(
+    dep_node_handles_by_node: Mapping[NodeHandle, Sequence[NodeHandle]],
+    curr_node_handle: NodeHandle,
+    graph_output_name: Optional[str] = None,
+):
```
A return type annotation would be helpful here.

clairelin135 replied:
Thanks, added!

sryza commented on python_modules/dagster/dagster/core/definitions/assets_info.py:
```diff
@@ -64,6 +67,188 @@ def __new__(
+    dep_nodes_by_asset_key: Dict[AssetKey, List[NodeHandle]] = {}
+
+    for node_handle, assets_defs in assets_defs_by_node_handle.items():
+        dep_node_handles_by_node: Dict[
```
Could this be taken out of the for loop? Seems like we could avoid redoing work by sharing these across all the assets, given that it's all the same graph?

clairelin135 replied:
I think this dictionary should stay here--each `node_handle` in this outer loop is a top level op/graph in the asset job, and there won't be any shared node handles that are nested in different top level `node_handle`s. So this dictionary exists to memoize traversals for every single asset output for the given top level `node_handle`.

sryza reviewed (commented):
Way way more readable! Thanks for these changes. I left one question, then I think this should be good to go.

OwenKephart reviewed (approved):
takes some time to understand what's going on here, but overall looks good. definitely a tough problem :)

you'll probably want to rebase off of owen/asset_op_interop_graphs to minimize conflicts. I think the most annoying thing you'll need to update on that rebase is that input_names_by_asset_key on AssetsDefinition got changed to asset_keys_by_input_name, so your tests will need to be updated.

clairelin135 commented on python_modules/dagster/dagster/core/definitions/assets_info.py:
```diff
@@ -64,6 +67,188 @@ def __new__(
+            )
+            outputs_by_graph_handle[curr_node_handle] = {
+                mapping.definition.name: (
+                    mapping.maps_from.output_name,
```
right right, I've changed all of these tuples of output_name, node_handle to NodeOutputHandles. thanks--wasn't aware that this object existed
